substrate/client/network/sync/src/state_strategy.rs

```diff
@@ -0,0 +1,321 @@
+		}
+	}
+
+	fn on_state_response_inner(
+		&mut self,
+		peer_id: PeerId,
+		response: OpaqueStateResponse,
+	) -> Result<(), BadPeer> {
+		if let Some(peer) = self.peers.get_mut(&peer_id) {
+			peer.state = PeerState::Available;
+		}
+
+		let response: Box<StateResponse> = response.0.downcast().map_err(|_error| {
+			error!(
+				target: LOG_TARGET,
+				"Failed to downcast opaque state response, this is an implementation bug."
+			);
+
+			BadPeer(peer_id, rep::BAD_RESPONSE)
+		})?;
+
+		debug!(
+			target: LOG_TARGET,
+			"Importing state data from {} with {} keys, {} proof nodes.",
+			peer_id,
+			response.entries.len(),
+			response.proof.len(),
+		);
+
+		match self.state_sync.import(*response) {
+			ImportResult::Import(hash, header, state, body, justifications) => {
+				let origin = BlockOrigin::NetworkInitialSync;
+				let block = IncomingBlock {
+					hash,
+					header: Some(header),
+					body,
+					indexed_body: None,
+					justifications,
+					origin: None,
+					allow_missing_state: true,
+					import_existing: true,
+					skip_execution: true,
+					state: Some(state),
+				};
+				debug!(target: LOG_TARGET, "State download is complete. Import is queued");
+				self.actions
+					.push(StateStrategyAction::ImportBlocks { origin, blocks: vec![block] });
+				Ok(())
+			},
+			ImportResult::Continue => Ok(()),
+			ImportResult::BadResponse => {
+				debug!(target: LOG_TARGET, "Bad state data received from {peer_id}");
+				Err(BadPeer(peer_id, rep::BAD_STATE))
+			},
+		}
+	}
+
+	/// A batch of blocks have been processed, with or without errors.
+	///
+	/// Normally this should be called when target block with state is imported.
+	pub fn on_blocks_processed(
+		&mut self,
+		imported: usize,
+		count: usize,
+		results: Vec<(Result<BlockImportStatus<NumberFor<B>>, BlockImportError>, B::Hash)>,
+	) {
+		trace!(target: LOG_TARGET, "State sync: imported {imported} of {count}.");
+
+		let results = results
+			.into_iter()
+			.filter_map(|(result, hash)| {
+				if hash == self.state_sync.target() {
+					Some(result)
+				} else {
+					debug!(
+						target: LOG_TARGET,
+						"Unexpected block processed: {hash} with result {result:?}.",
+					);
+					None
+				}
+			})
+			.collect::<Vec<_>>();
+
+		if !results.is_empty() {
+			// We processed the target block
+			results.iter().filter_map(|result| result.as_ref().err()).for_each(|e| {
+				error!(
+					target: LOG_TARGET,
+					"Failed to import target block with state: {e:?}."
+				);
+			});
+			match results.into_iter().any(|result| result.is_ok()) {
+				true => {
+					info!(
+						target: LOG_TARGET,
+						"State sync is complete ({} MiB), continuing with block sync.",
+						self.state_sync.progress().size / (1024 * 1024),
+					);
+				},
+				false => {
+					error!(
+						target: LOG_TARGET,
+						"State sync failed. Falling back to full sync.",
+					);
+					// TODO: Test this scenario.
+					//       Make sure fail-over doesn't render useless warp sync tests.
```

Comment:
I'll leave a comment here to ensure I don't forget to address the TODO. With `ChainSync` started even if warp sync has failed, the `warp_sync` test in `sc-network-test` might not detect the warp sync issues, because the synchronization will be performed by `ChainSync` anyway.

Reply:
`sync::warp_sync` tests that we don't have a state for block 1 (target = 64), i.e. the synchronization wasn't done by full sync. Therefore not an issue.